examples/nodejs/storage/user-program.ts

```diff
@@ -0,0 +1,98 @@
+
+      // To request back as a concrete type means the developer knows the type.
+      // If the developer is right, they get the value. And if they are wrong,
+      // we have two choices:
+      // 1. return undefined, or
+      // 2. throw an exception.
+      //
+      // Returning undefined is a silent failure. The developer may not know
+      // that the value was not what they expected. They may also confuse
+      // the undefind with the value not in the store.
+      //
+      // Throwing an exception is a loud failure. The developer will know
+      // immediately they were wrong. But not all developers want to
+      // handle exceptions.
+      //
+      // We can make it easy for the developer to exhaustively check the type
+      // and handle the cases they care about as follows:
+      switch (getResponse.type) {
+        case StoreValueType.STRING:
+          console.log('The value is a string:', getResponse.tryGetValueString());
+          break;
+        case StoreValueType.INTEGER:
+          console.log('The value is an integer:', getResponse.tryGetValueInteger());
+          break;
+        case StoreValueType.FLOAT:
+          console.log('The value is a float:', getResponse.tryGetValueFloat());
+          break;
+        case StoreValueType.BOOLEAN:
+          console.log('The value is a boolean:', getResponse.tryGetValueBoolean());
+          break;
+        default:
+          console.error('Unknown type');
+      }
+      // An alternative to this switch is:
+      // - StoreGet.Success has a `value` field which is of type `StoreValue`
+      // `StoreValue` is a base class with derived types, each with a `value` field.
+      // - StoreStringValue with a string-valued `value` field,
+      // - StoreIntegerValue with an integer-valued `value` field,
+      // - StoreFloatValue, etc.
+      // In that world the developer can use more instanceof checks:
+      // if (getResponse instanceof StoreGet.Success) {
+      //    const storeValue = getResponse.value;
+      //    if (storeValue instanceof StoreStringValue) {
+      //      console.log('The value is a string:', storeValue.value);
+      //    } else if (storeValue instanceof StoreIntegerValue) {
+      //      console.log('The value is an integer:', storeValue.value);
+      //    } else if (storeValue instanceof StoreFloatValue) {
+      //      // ...
+      //    } else {
+      //      console.error('Unknown type');
+      //    }
+      // }
+      //
+      // I feel this pattern makes sense in a language with first-class pattern matching, such as
+      // Rust, Scala, Kotlin, etc.
+      // But in JavaScript, I feel the switch is at the right level of complexity.
+      // This introduces an extra level of indirection where the dev has to access the `Success`
+      // value and then the `value` field of the `StoreValue` object.
+    } else if (getResponse instanceof StoreGet.Error) {
```

Comment:
Since we are most likely down to a maximum of 2 different types available for any given response (Success/Error), I'm less certain now that it makes sense to model the errors as a return type. *Personally* I would still prefer that, but it's harder to make a case against the language idioms when it's only two types instead of 3.

In go/Rust, there are first-class patterns for this but we don't have that here. I assume if you play with the DDB, redis, or mongo node.js libs they will just throw and not introduce this indirection.

It'd be compelling to be able to get rid of the entire `if` block here for happy path examples, though I realize that would basically imply that `try/catch` was mandatory for production-quality code. It'd also have big implications for consistency of UX if it is shipped in the same package as the CacheClient :)

If nothing else you could move the `.tryGetValue*` methods up to the parent class so that they could be accessed directly on the response without having to do the instance check first?

Reply:
> Since we are most likely down to a maximum of 2 different types available for any given response (Success/Error), I'm less certain now that it makes sense to model the errors as a return type. Personally I would still prefer that, but it's harder to make a case against the language idioms when it's only two types instead of 3.

One thing I punted on exploring here was how to surface the `not found` case. In the cache case we have:
- hit
- miss
- error, including cache not found

Whereas in this PR we have:
- success (hit)
- error, which includes both store not found, key not found, and the other gamut of errors.

The latter here is consistent with HTTP and gRPC response codes. We have a 200 for OK/resource found and a 404 for not found. So I would expect to be writing code like:
```typescript
if (getResponse instanceof StoreGet.Success) {
// ...
} else if (getResponse instanceof StoreGet.Error) {
   switch (getResponse.errorCode) {
       // auto gen'd based off of an enum with the most common at the top
      case StoreErrorCode.KEY_NOT_FOUND:
         //...
      case StoreErrorCode.STORE_NOT_FOUND:
        // ...
// ...
}
```

Reply:
re: switching on error codes, yes, agreed. But I think my bigger question is whether it even makes sense to have to do the instanceof check when we know there are only 2 possible cases.

To say it another way... I think before, when there were 3 cases, we argued that users were going to need an `if` block no matter what, so it made more sense to push the error into a 3rd branch of the if statement rather than forcing them to wrap the `if` with a `try/catch`.

But in this case, users don't necessarily need the `if` block at all if they only care about Success. and adding a try/catch around the API call is no longer (necessarily) any more unwieldy than adding the `if` you're using here.

So I'm asking whether it should just be `const value = await client.get(storeName, key);`

Reply:
I feel like `const value = await client.get(storeName, key);` is broadly more idiomatic and less cluttered. I think users would expect a missing value to be `undefined` though (ie the return value is optional), in which case they can branch on `undefined` or assume it's there.

That said, if we do that, then we lose forward compatibility for this API.

Reply:
In the CacheClient we support `.value()` on the response for those who want to short-circuit the pattern matching / type checking stuff, maybe that's sufficient here too